packages/backend/src/services/ProjectService/ProjectService.ts

```diff
@@ -122,32 +130,60 @@ export class ProjectService {
         this.jobModel = jobModel;
         this.emailClient = emailClient;
         this.spaceModel = spaceModel;
+        this.sshKeyPairModel = sshKeyPairModel;
     }
 
-    private async _getWarehouseClient(
-        projectUuid: string,
-    ): Promise<WarehouseClient> {
+    private async _resolveWarehouseClientSshKeys<
+        T extends { warehouseConnection: CreateWarehouseCredentials },
+    >(args: T): Promise<T> {
+        if (
+            (args.warehouseConnection.type === WarehouseTypes.REDSHIFT ||
+                args.warehouseConnection.type === WarehouseTypes.POSTGRES) &&
+            args.warehouseConnection.useSshTunnel
+        ) {
+            const publicKey = args.warehouseConnection.sshTunnelPublicKey || '';
+            const { privateKey } = await this.sshKeyPairModel.get(publicKey);
+            return {
+                ...args,
+                warehouseConnection: {
+                    ...args.warehouseConnection,
+                    sshTunnelPrivateKey: privateKey,
+                },
+            };
+        }
+        return args;
+    }
+
+    private async _getWarehouseClient(projectUuid: string): Promise<{
+        warehouseClient: WarehouseClient;
+        sshTunnel: SshTunnel<CreateWarehouseCredentials>;
+    }> {
         // Always load the latest credentials from the database
         const credentials =
             await this.projectModel.getWarehouseCredentialsForProject(
                 projectUuid,
             );
-        // Check cache for existing client
+        // Setup SSH tunnel for client (user needs to close this)
+        const sshTunnel = new SshTunnel(credentials);
```

Comment:
I feel like this tunnel should be optional, so people without tunnes will not run any of this, like 

```
if (sshCredentials) {
const sshTunnel = new SshTunnel(credentials);
sshTunnel.connect();
} else {
client.connect(); 
}